File: hello-agents-main/Co-creation-projects/xujikai-SentenceExpandAgent/frontend/src/api/expand.ts
/**
 * 英语句子扩写智能体 - API 封装
 * 包含 REST API 和 SSE 流式接口
 */

import type {
  StartRequest,
  SubmitRequest,
  AgentResponse,
  SessionState,
  SSEEvent
} from '../types/expand';

// API 基础 URL
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';

/**
 * 开始新的扩写会话
 */
export async function startSession(request: StartRequest): Promise<AgentResponse> {
  const response = await fetch(`${API_BASE_URL}/api/session/start`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`Failed to start session: ${response.statusText}`);
  }

  return response.json();
}

/**
 * 提交用户扩写句子（手动模式）
 */
export async function submitSentence(request: SubmitRequest): Promise<AgentResponse> {
  const response = await fetch(`${API_BASE_URL}/api/session/submit`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`Failed to submit sentence: ${response.statusText}`);
  }

  return response.json();
}

/**
 * 获取会话完整状态
 */
export async function getSession(sessionId: string): Promise<SessionState> {
  const response = await fetch(`${API_BASE_URL}/api/session/${sessionId}`);

  if (!response.ok) {
    throw new Error(`Failed to get session: ${response.statusText}`);
  }

  return response.json();
}

/**
 * SSE 流式自动模式
 * @param sessionId 会话 ID
 * @param onMessage 接收到消息消息时的回调
 * @param onError 发生错误时的回调
 * @param onComplete 完成时的回调
 * @returns 清理函数，用于关闭 EventSource
 */
export function subscribeAutoMode(
  sessionId: string,
  onMessage: (event: SSEEvent) => void,
  onError?: (error: Error) => void,
  onComplete?: () => void
): () => void {
  const eventSource = new EventSource(
    `${API_BASE_URL}/api/session/${sessionId}/auto`
  );

  // 处理普通消息
  eventSource.onmessage = (event) => {
    try {
      const data = JSON.parse(event.data);
      onMessage({
        type: data.type || 'question',
        data: data.data,
      });
    } catch (error) {
      console.error('Failed to parse SSE message:', error);
    }
  };

  // 处理完成事件
  eventSource.addEventListener('done', () => {
    onComplete?.();
    eventSource.close();
  });

  // 处理错误事件
  eventSource.addEventListener('error', (event) => {
    try {
      const data = JSON.parse(event.data);
      onMessage({
        type: 'error',
        error: data.error,
      });
    } catch (error) {
      console.error('Failed to parse error event:', error);
    }
    onError?.(new Error('Auto mode error'));
  });

  // 处理连接错误
  eventSource.onerror = (error) => {
    console.error('SSE connection error:', error);
    onError?.(new Error('SSE connection failed'));
    eventSource.close();
  };

  // 返回清理函数
  return () => {
    eventSource.close();
  };
}

File: hello-agents-main/Co-creation-projects/xujikai-SentenceExpandAgent/backend/src/routers/expand.py
"""
FastAPI 路由层 - 英语句子扩写智能体
"""
from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from typing import AsyncGenerator
import sys
import os

# 添加 backend 目录到 Python 路径
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from models.entities import (
    StartRequest,
    SubmitRequest,
    AgentResponse,
    SessionState
)
from services.session_store import get_session_store
from agents.orchestrator import get_orchestrator
from agents.auto_mode_agent import get_auto_mode

router = APIRouter(prefix="/api", tags=["expand"])


@router.post("/session/start", response_model=AgentResponse)
async def start_session(request: StartRequest) -> AgentResponse:
    """
    创建新会话，返回第一阶段提问
    
    Args:
        request: 开始会话请求，包含种子句和模式
        
    Returns:
        AgentResponse: 智能体响应
    """
    # 获取会话存储
    session_store = get_session_store()
    
    # 创建会话
    session = session_store.create_session(
        seed_sentence=request.seed_sentence,
        mode=request.mode
    )
    
    # 获取 Orchestrator
    orchestrator = get_orchestrator()
    
    # 开始会话
    response = orchestrator.start_session(session)
    
    return response


@router.post("/session/submit", response_model=AgentResponse)
async def submit_sentence(request: SubmitRequest) -> AgentResponse:
    """
    提交用户扩写句子，返回点评和下一阶段提问（手动模式）
    
    Args:
        request: 提交请求，包含会话 ID 和用户句子
        
    Returns:
        AgentResponse: 智能体响应
    """
    # 获取会话存储
    session_store = get_session_store()
    
    # 获取会话
    session = session_store.get_session(request.session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    
    # 获取 Orchestrator
    orchestrator = get_orchestrator()
    
    # 处理用户输入
    response = orchestrator.process_user_input(
        session_state=session,
        user_sentence=request.user_sentence
    )
    
    # 更新会话
    session_store.update_session(session)
    
    return response


@router.get("/session/{session_id}/auto")
async def auto_mode_stream(session_id: str) -> StreamingResponse:
    """
    SSE 流式推送三轮自动演示
    
    Args:
        session_id: 会话 ID
        
    Returns:
        StreamingResponse: SSE 流式响应
    """
    # 获取会话存储
    session_store = get_session_store()
    
    # 获取会话
    session = session_store.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    
    # 获取 AutoModeAgent
    auto_mode_agent = get_auto_mode()
    
    # 生成流式响应
    async def event_generator() -> AsyncGenerator[str, None]:
        import json
        try:
            # 使用流式运行
            async for event in auto_mode_agent.run_auto_mode_stream(session.seed_sentence):
                yield f"data: {json.dumps(event, ensure_ascii=False)}\n\n"
            # 发送结束事件
            yield "event: done\ndata: {}\n\n"
        except Exception as e:
            yield f"event: error\ndata: {json.dumps({'detail': str(e), 'type': type(e).__name__}, ensure_ascii=False)}\n\n"
    
    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no"
        }
    )


@router.get("/session/{session_id}", response_model=SessionState)
async def get_session(session_id: str) -> SessionState:
    """
    获取当前会话完整状态
    
    Args:
        session_id: 会话 ID
        
    Returns:
        SessionState: 会话状态
    """
    # 获取会话存储
    session_store = get_session_store()
    
    # 获取会话
    session = session_store.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    
    return session
